datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+            fulfilled=fulfilled, recursive=recursive, recursion_limit=recursion_limit,
+            bottomup=bottomup,
+            result_xfm='paths'
+        )
+
+        if subdatasets_only:
+            datasets_it = subdatasets_it
+        else:
+            if bottomup:
+                datasets_it = chain(subdatasets_it, [refds.path])
+            else:
+                datasets_it = chain([refds.path], subdatasets_it)
+
+        if not python:
+            protocol = NoCapture if passthrough else StdOutErrCapture
+
+        def run_cmd(dspath):
+            ds = Dataset(dspath)
+            status_rec = get_status_dict(
+                'foreach',
+                ds=ds,
+                path=ds.path,
+                command=cmd
+            )
+            if not ds.is_installed():
+                yield dict(
+                    status_rec,
+                    status="impossible",
+                    message="not installed"
+                )
+                return
+            # For consistent environment (Python) and formatting (command) similar to `run` one
+            # But for Python command we provide actual ds and refds not paths
+            placeholders = dict(
+                pwd=pwd,
+                ds=ds if python else ds.path,
+                refds=refds if python else refds.path,
+                # Check if the command contains "tmpdir" to avoid creating an
+                # unnecessary temporary directory in most but not all cases.
+                # Note: different from 'run' - not wrapping match within {} and doing str
+                tmpdir=mkdtemp(prefix="datalad-run-") if "tmpdir" in str(cmd) else "")
+            try:
+                if python:
+                    status_rec['results'] = eval(cmd[0], placeholders)
+                else:
+                    try:
+                        cmd_expanded = format_command(cmd, **placeholders)
+                    except KeyError as exc:
+                        yield dict(
+                            status_rec,
+                            status='impossible',
+                            message=('command has an unrecognized placeholder: %s', exc))
+                        return
+                    # TODO: avoid use of _git_runner
+                    out = ds.repo._git_runner.run(cmd_expanded, protocol=protocol)
+                    if not passthrough:
+                        status_rec.update(out)
+                        # provide some feedback to user in default rendering
+                        if any(out.values()):
+                            status_rec['message'] = shortened_repr(out, 100)
+                status_rec['status'] = 'ok'
+            except Exception as exc:
+                # TODO: option to not swallow but reraise!
+                status_rec['status'] = 'error'
+                # TODO: there must be a better place for it since this one is not
+                # output by -f json_pp ... a feature or a bug???
```

Comment:
It's intentional:

https://github.com/datalad/datalad/blob/3f5b53c0de72fa826c0cb4d2ef0acc86dd19aa28/datalad/interface/utils.py#L633-L635

`get_status_dict` would put this in "exception_traceback".

Reply:
but it would leave user wondering on what has actually happened 
<details>
<summary>since no message would be provided</summary> 

```shell
$> datalad -f json_pp foreach -d ~/datalad/openneuro --fulfilled true --bottomup touch /dev/123 | head -n 14
{
  "action": "foreach",
  "command": [
    "touch",
    "/dev/123"
  ],
  "exception_traceback": "[foreach.py:run_cmd:244,cmd.py:run:408]",
  "path": "/home/yoh/datalad/openneuro/ds000001",
  "status": "error",
  "type": "dataset"
}
{
  "action": "foreach",
  "command": [
[ERROR] [Errno 32] Broken pipe [ansitowin32.py:write_plain_text:196] (BrokenPipeError) 
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
BrokenPipeError: [Errno 32] Broken pipe
(dev3) 1 67232 [1].....................................:Mon 22 Mar 2021 07:22:07 PM EDT:.
(git-annex)lena:~/datalad[master].datalad
$> datalad  foreach -d ~/datalad/openneuro --fulfilled true --bottomup touch /dev/123         
foreach(error): /home/yoh/datalad/openneuro/ds000001 (dataset)                                            
foreach(error): /home/yoh/datalad/openneuro/ds000002 (dataset)   
...
```
</details>

so I guess I will use `get_status_dict` but also pass `str(exc)` as message

Reply:
gh-5536 proposes to retain the message in the `json{,_pp}` output.

Reply:
good.